src/lib.rs

```diff
@@ -39,12 +44,222 @@ pub enum Error {
+    Overlap(Range, Range),
     /// A node state can be changed just from Free to Allocated, other transitions
     /// are not valid.
     #[error("Invalid state transition for node {0:?} from {1:?} to NodeState::Free")]
-    InvalidStateTransition(allocation_engine::Range, allocation_engine::NodeState),
+    InvalidStateTransition(Range, NodeState),
+    /// Address is unaligned
+    #[error("The address is not aligned.")]
+    UnalignedAddress,
 }
 
 /// Wrapper over std::result::Result
 pub type Result<T> = result::Result<T, Error>;
+
+/// A closed interval range [start, end] used to describe a memory slot that
+/// will be assigned to a device by the VMM. This structure represents the key
+/// of the Node object in the interval tree implementation.
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Hash, Ord, Debug)]
+pub struct Range {
+    /// Lower boundary of the interval.
+    start: u64,
+    /// Upper boundary of the interval.
+    end: u64,
+}
+
+#[allow(clippy::len_without_is_empty)]
+impl Range {
+    /// Create a new Range object.
+    pub fn new(start: u64, end: u64) -> Result<Self> {
+        if start > end {
+            return Err(Error::InvalidRange(start, end));
+        }
+        Ok(Range { start, end })
+    }
+
+    /// Get length of the range.
+    pub fn len(&self) -> u64 {
+        self.end - self.start + 1
+    }
+
+    /// Check whether two Range objects overlap with each other.
+    pub fn overlaps(&self, other: &Range) -> bool {
+        max(self.start, other.start) <= min(self.end, other.end)
+    }
+
+    /// Check whether the key is fully covered.
+    pub fn contains(&self, other: &Range) -> bool {
+        self.start <= other.start && self.end >= other.end
+    }
+
+    /// Get the lower boundary of the range.
+    pub fn start(&self) -> u64 {
+        self.start
+    }
+
+    /// Get the upper boundary of the range.
+    pub fn end(&self) -> u64 {
+        self.end
+    }
+
+    /// Set the lower boundary of the range.
+    pub fn set_start(&mut self, start: u64) {
+        self.start = start;
+    }
+
+    /// Set the upper boundary of the range.
+    pub fn set_end(&mut self, end: u64) {
+        self.end = end;
+    }
+}
+
+/// Struct to describe resource allocation constraints.
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub struct Constraint {
+    /// Size to allocate.
+    size: u64,
+    /// Range where the allocated resource will be placed.
+    desired_range: Range,
+    /// Alignment for the allocated resource.
+    align: u64,
+    /// Resource allocation policy.
+    policy: AllocPolicy,
+}
+
+impl Constraint {
+    /// Create a new constraint object with default settings.
+    pub fn new(size: u64) -> Result<Self> {
+        Ok(Constraint {
+            size,
+            desired_range: Range::new(0, std::u64::MAX)?,
+            align: 4,
+            policy: AllocPolicy::default(),
+        })
+    }
+
+    /// Set the min constraint.
+    pub fn desired_range(mut self, range: Range) -> Self {
```

Comment:
There seems to be a mismatch between the name of the function and the doc comment. Which one is correct?